src/main/java/seedu/taskitty/model/ModelManager.java

```diff
@@ -105,19 +107,16 @@ public synchronized void addTask(Task task) throws UniqueTaskList.DuplicateTaskE
         indicateTaskManagerChanged();
     }
     
-    public synchronized String undo() {
-        resetData(historyTaskManagers.pop());   
-        updateFilteredTaskList(historyPredicates.pop());
-        return historyCommands.pop();
-    }
-    
-    //@@author A0135793W
-    private void updateFilters() {
-        filteredTodos = new FilteredList<>(taskManager.getFilteredTodos());
-        filteredDeadlines = new FilteredList<>(taskManager.getFilteredDeadlines());
-        filteredEvents = new FilteredList<>(taskManager.getFilteredEvents());     
+    //@@ author A0139052L
+    public synchronized String undo() throws NoPreviousCommandException {
+        if (noMorePreviousCommand()) {            
+            throw new NoPreviousCommandException(Messages.MESSAGE_NO_PREVIOUS_COMMANDS);
+        }
+        assert !historyPredicates.isEmpty() && !historyTaskManagers.isEmpty();
```

Comment:
It first checks that there should be a previous command. If there exists previous commands, then there should be a TaskManager and Predicate also saved, which is why i assert there.